parser/src/parser.rs

```diff
@@ -10,23 +11,26 @@ use crate::aspects::test::TestAspect;
 use crate::aspects::var_declaration::VarDeclarationAspect;
 use crate::ast::Expr;
 use crate::cursor::ParserCursor;
-use crate::moves::{bin_op, eod, eox, next, of_types, spaces, MoveOperations, repeat, space};
+use crate::err::{ErrorContext, ParseError, ParseErrorKind};
+use crate::moves::{bin_op, eod, eox, next, of_types, repeat, space, spaces, MoveOperations};
 
-pub type ParseResult<T> = Result<T, ParseError>;
-
-/// An error that occurs during parsing.
-#[derive(Debug, PartialEq)]
-pub struct ParseError {
-    pub message: String,
-    //pub actual: Token<'a>,
-}
+pub(crate) type ParseResult<T> = Result<T, ParseError>;
 
 /// A parser for the Moshell scripting language.
 pub(crate) struct Parser<'a> {
     pub(crate) cursor: ParserCursor<'a>,
+    pub(crate) source: Source<'a>,
 }
 
 impl<'a> Parser<'a> {
+    /// Creates a new parser from a defined source.
+    pub(crate) fn new(source: Source<'a>) -> Self {
+        Self {
+            cursor: ParserCursor::new_with_source(lex(source.source), source.source),
+            source,
+        }
+    }
+
     /// Parses input tokens into an abstract syntax tree representation.
     pub fn parse(&mut self) -> ParseResult<Vec<Expr<'a>>> {
```

Comment:
Kinda weird that the `parse` method still return a result whose Error is the last error encountered. This results in error waste as all the error _excepted the last_ will not be part of the `ParseReport` structure. 
If this parser parses a file, only the last error would get reported.